rllib/evaluation/worker_set.py

```diff
@@ -128,91 +130,201 @@ def __init__(
+            trainer_config: Optional dict that extends the common config of
+                the Algorithm class.
+            num_workers: Number of remote rollout workers to create.
+            local_worker: Whether to create a local (non @ray.remote) worker
+                in the returned set as well (default: True). If `num_workers`
+                is 0, always create a local worker.
+        """
+        # Force a local worker if num_workers == 0 (no remote workers).
+        # Otherwise, this WorkerSet would be empty.
+        self._local_worker = None
+        if num_workers == 0:
+            local_worker = True
+        self._local_config = config.copy(copy_frozen=False).framework(
+            tf_session_args=config.local_tf_session_args
+        )
+
+        if config.input_ == "dataset":
+            # Create the set of dataset readers to be shared by all the
+            # rollout workers.
+            self._ds, self._ds_shards = get_dataset_and_shards(config, num_workers)
+        else:
+            self._ds = None
+            self._ds_shards = None
+
+        # Create a number of @ray.remote workers.
+        self.__worker_manager = FaultTolerantActorManager()
+        self.add_workers(
+            num_workers,
+            validate=config.validate_workers_after_construction,
+        )
+
+        # If num_workers > 0 and we don't have an env on the local worker,
+        # get the observation- and action spaces for each policy from
+        # the first remote worker (which does have an env).
+        if (
+            local_worker
+            and self.__worker_manager.num_actors() > 0
+            and not config.create_env_on_local_worker
+            and (not config.observation_space or not config.action_space)
+        ):
+            spaces = self._get_spaces_from_remote_worker()
+        else:
+            spaces = None
+
+        # Create a local worker, if needed.
+        if local_worker:
+            self._local_worker = self._make_worker(
+                cls=RolloutWorker,
+                env_creator=self._env_creator,
+                validate_env=validate_env,
+                worker_index=0,
+                num_workers=num_workers,
+                config=self._local_config,
+                spaces=spaces,
             )
 
-            if config.input_ == "dataset":
-                # Create the set of dataset readers to be shared by all the
-                # rollout workers.
-                self._ds, self._ds_shards = get_dataset_and_shards(config, num_workers)
-            else:
-                self._ds = None
-                self._ds_shards = None
-
-            # Create a number of @ray.remote workers.
-            self._remote_workers = []
-            self.add_workers(
-                num_workers,
-                validate=config.validate_workers_after_construction,
+    def _get_spaces_from_remote_worker(self):
+        """Guess observation and action spaces from a remote worker.
```

Comment:
"Guess" -> "Infer". If the remote worker has these spaces, we would trust that information, I think.

Reply:
sure sure. I trust it too.